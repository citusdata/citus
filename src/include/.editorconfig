# Don't change generated input file
[citus_config.h.in]
insert_final_newline = unset
trim_trailing_whitespace = unset
indent_style = unset
indent_size = unset
