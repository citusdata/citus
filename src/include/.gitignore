/stamp-h
/stamp-ext-h
/citus_config.h
/citus_config.h.in~
/citus_version.h
/citus_version.h.in~
