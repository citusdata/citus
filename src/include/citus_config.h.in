/* src/include/citus_config.h.in.  Generated from configure.in by autoheader.  */


/*
 * citus_config.h.in is generated by autoconf/autoheader and
 * converted into citus_config.h by configure.  Include when code needs to
 * depend on determinations made by configure.
 *
 * Do not manually edit!
 */


/* Citus edition as a string */
#undef CITUS_EDITION

/* Extension version expected by this Citus build */
#undef CITUS_EXTENSIONVERSION

/* Citus major version as a string */
#undef CITUS_MAJORVERSION

/* Citus full name as a string */
#undef CITUS_NAME

/* Citus version as a string */
#undef CITUS_VERSION

/* Citus version as a number */
#undef CITUS_VERSION_NUM

/* A string containing the version number, platform, and C compiler */
#undef CITUS_VERSION_STR

/* Define to 1 if you have the <inttypes.h> header file. */
#undef HAVE_INTTYPES_H

/* Define to 1 if you have the <memory.h> header file. */
#undef HAVE_MEMORY_H

/* Define to 1 if you have the <stdint.h> header file. */
#undef HAVE_STDINT_H

/* Define to 1 if you have the <stdlib.h> header file. */
#undef HAVE_STDLIB_H

/* Define to 1 if you have the <strings.h> header file. */
#undef HAVE_STRINGS_H

/* Define to 1 if you have the <string.h> header file. */
#undef HAVE_STRING_H

/* Define to 1 if you have the <sys/stat.h> header file. */
#undef HAVE_SYS_STAT_H

/* Define to 1 if you have the <sys/types.h> header file. */
#undef HAVE_SYS_TYPES_H

/* Define to 1 if you have the <unistd.h> header file. */
#undef HAVE_UNISTD_H

/* Define to the address where bug reports for this package should be sent. */
#undef PACKAGE_BUGREPORT

/* Define to the full name of this package. */
#undef PACKAGE_NAME

/* Define to the full name and version of this package. */
#undef PACKAGE_STRING

/* Define to the one symbol short name of this package. */
#undef PACKAGE_TARNAME

/* Define to the home page for this package. */
#undef PACKAGE_URL

/* Define to the version of this package. */
#undef PACKAGE_VERSION

/* Base URL for statistics collection and update checks */
#undef REPORTS_BASE_URL

/* The size of `void *', as computed by sizeof. */
#undef SIZEOF_VOID_P

/* Define to 1 if you have the ANSI C header files. */
#undef STDC_HEADERS
