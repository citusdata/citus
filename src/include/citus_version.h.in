/* This file is created manually */

/* Citus full name as a string */
#undef CITUS_NAME

/* Citus edition as a string */
#undef CITUS_EDITION

/* Extension version expected by this Citus build */
#undef CITUS_EXTENSIONVERSION

/* Citus major version as a string */
#undef CITUS_MAJORVERSION

/* Citus version as a string */
#undef CITUS_VERSION

/* Citus version as a number */
#undef CITUS_VERSION_NUM

/* A string containing the version number, platform, and C compiler */
#undef CITUS_VERSION_STR

/* Define to 1 if you have the `curl' library (-lcurl). */
#undef HAVE_LIBCURL

/* Define to 1 if you have the `liblz4' library (-llz4). */
#undef HAVE_LIBLZ4

/* Define to 1 if you have the `libzstd' library (-lzstd). */
#undef HAVE_LIBZSTD

/* Base URL for statistics collection and update checks */
#undef REPORTS_BASE_URL

/* columnar table access method capability */
#undef HAS_TABLEAM
