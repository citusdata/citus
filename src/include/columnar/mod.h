/*-------------------------------------------------------------------------
 *
 * mod.h
 *
 * Type and function declarations for columnar
 *
 * Copyright (c) 2016, Citus Data, Inc.
 *
 * $Id$
 *
 *-------------------------------------------------------------------------
 */

#ifndef MOD_H
#define MOD_H

/* Function declarations for extension loading and unloading */
extern void columnar_init(void);
extern void columnar_fini(void);

#endif /* MOD_H */
