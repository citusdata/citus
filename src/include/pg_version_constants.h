/*-------------------------------------------------------------------------
 *
 * pg_version_constants.h
 *	   pg version related constants.
 *
 * Copyright (c) Citus Data, Inc.
 *
 *-------------------------------------------------------------------------
 */

#ifndef PG_VERSION_CONSTANTS
#define PG_VERSION_CONSTANTS

#define PG_VERSION_15 150000
#define PG_VERSION_16 160000
#define PG_VERSION_17 170000
#define PG_VERSION_18 180000

#endif   /* PG_VERSION_CONSTANTS */
