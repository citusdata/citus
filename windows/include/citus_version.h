/* Citus full name as a string */
#define CITUS_NAME "Citus"

/* Citus edition as a string */
#define CITUS_EDITION "community"

/* Extension version expected by this Citus build */
#define CITUS_EXTENSIONVERSION "7.3-1"

/* Citus major version as a string */
#define CITUS_MAJORVERSION "7.3"

/* Citus version as a string */
#define CITUS_VERSION "7.3devel"

/* Citus version as a number */
#define CITUS_VERSION_NUM 70300

/* A string containing the version number, platform, and C compiler */
#define CITUS_VERSION_STR "Citus 7.3devel on x86_64-windows, compiled by Visual Studio"

/* Define to 1 if you have the `curl' library (-lcurl). */
#define HAVE_LIBCURL 0

/* Base URL for statistics collection and update checks */
#define REPORTS_BASE_URL "https://reports.citusdata.com"
