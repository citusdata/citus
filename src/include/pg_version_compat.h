/*-------------------------------------------------------------------------
 *
 * pg_version_compat.h
 *	  Compatibility macros for writing code agnostic to PostgreSQL versions
 *
 * Copyright (c) Citus Data, Inc.
 *
 *-------------------------------------------------------------------------
 */

#ifndef PG_VERSION_COMPAT_H
#define PG_VERSION_COMPAT_H

#include "distributed/pg_version_constants.h"

#if PG_VERSION_NUM >= PG_VERSION_15
#define ProcessCompletedNotifies()
#define RelationCreateStorage_compat(a, b, c) RelationCreateStorage(a, b, c)
#define parse_analyze_varparams_compat(a, b, c, d, e) parse_analyze_varparams(a, b, c, d, \
																			  e)
#define CREATE_SEQUENCE_COMMAND \
	"CREATE %sSEQUENCE IF NOT EXISTS %s AS %s INCREMENT BY " INT64_FORMAT \
	" MINVALUE " INT64_FORMAT " MAXVALUE " INT64_FORMAT \
	" START WITH " INT64_FORMAT " CACHE " INT64_FORMAT " %sCYCLE"
#else

#include "nodes/value.h"
#include "storage/smgr.h"
#include "utils/int8.h"
#include "utils/rel.h"

typedef Value String;

#ifdef HAVE_LONG_INT_64
#define strtoi64(str, endptr, base) ((int64) strtol(str, endptr, base))
#define strtou64(str, endptr, base) ((uint64) strtoul(str, endptr, base))
#else
#define strtoi64(str, endptr, base) ((int64) strtoll(str, endptr, base))
#define strtou64(str, endptr, base) ((uint64) strtoull(str, endptr, base))
#endif
#define RelationCreateStorage_compat(a, b, c) RelationCreateStorage(a, b)
#define parse_analyze_varparams_compat(a, b, c, d, e) parse_analyze_varparams(a, b, c, d)
#define pgstat_init_relation(r) pgstat_initstats(r)
#define pg_analyze_and_rewrite_fixedparams(a, b, c, d, e) pg_analyze_and_rewrite(a, b, c, \
																				 d, e)
#define boolVal(v) intVal(v)
#define makeBoolean(val) makeInteger(val)

static inline int64
pg_strtoint64(char *s)
{
	int64 result;
	(void) scanint8(s, false, &result);
	return result;
}


static inline SMgrRelation
RelationGetSmgr(Relation rel)
{
	if (unlikely(rel->rd_smgr == NULL))
	{
		smgrsetowner(&(rel->rd_smgr), smgropen(rel->rd_node, rel->rd_backend));
	}
	return rel->rd_smgr;
}


#define CREATE_SEQUENCE_COMMAND \
	"CREATE SEQUENCE IF NOT EXISTS %s AS %s INCREMENT BY " INT64_FORMAT \
	" MINVALUE " INT64_FORMAT " MAXVALUE " INT64_FORMAT \
	" START WITH " INT64_FORMAT " CACHE " INT64_FORMAT " %sCYCLE"

#endif

#if PG_VERSION_NUM >= PG_VERSION_14
#define AlterTableStmtObjType_compat(a) ((a)->objtype)
#define getObjectTypeDescription_compat(a, b) getObjectTypeDescription(a, b)
#define getObjectIdentity_compat(a, b) getObjectIdentity(a, b)

/* for MemoryContextMethods->stats */
#define stats_compat(a, b, c, d, e) stats(a, b, c, d, e)
#define FuncnameGetCandidates_compat(a, b, c, d, e, f, g) \
	FuncnameGetCandidates(a, b, c, d, e, f, g)
#define expand_function_arguments_compat(a, b, c, d) expand_function_arguments(a, b, c, d)
#define BeginCopyFrom_compat(a, b, c, d, e, f, g, h) BeginCopyFrom(a, b, c, d, e, f, g, h)
#define standard_ProcessUtility_compat(a, b, c, d, e, f, g, h) \
	standard_ProcessUtility(a, b, c, d, e, f, g, h)
#define ProcessUtility_compat(a, b, c, d, e, f, g, h) \
	ProcessUtility(a, b, c, d, e, f, g, h)
#define PrevProcessUtility_compat(a, b, c, d, e, f, g, h) \
	PrevProcessUtility(a, b, c, d, e, f, g, h)
#define SetTuplestoreDestReceiverParams_compat(a, b, c, d, e, f) \
	SetTuplestoreDestReceiverParams(a, b, c, d, e, f)
#define pgproc_statusflags_compat(pgproc) ((pgproc)->statusFlags)
#define get_partition_parent_compat(a, b) get_partition_parent(a, b)
#define RelationGetPartitionDesc_compat(a, b) RelationGetPartitionDesc(a, b)
#define make_simple_restrictinfo_compat(a, b) make_simple_restrictinfo(a, b)
#define pull_varnos_compat(a, b) pull_varnos(a, b)
#else
#define AlterTableStmtObjType_compat(a) ((a)->relkind)
#define F_NEXTVAL F_NEXTVAL_OID
#define ROLE_PG_MONITOR DEFAULT_ROLE_MONITOR
#define PROC_WAIT_STATUS_WAITING STATUS_WAITING
#define getObjectTypeDescription_compat(a, b) getObjectTypeDescription(a)
#define getObjectIdentity_compat(a, b) getObjectIdentity(a)

/* for MemoryContextMethods->stats */
#define stats_compat(a, b, c, d, e) stats(a, b, c, d)
#define FuncnameGetCandidates_compat(a, b, c, d, e, f, g) \
	FuncnameGetCandidates(a, b, c, d, e, g)
#define expand_function_arguments_compat(a, b, c, d) expand_function_arguments(a, c, d)
#define VacOptValue VacOptTernaryValue
#define VACOPTVALUE_UNSPECIFIED VACOPT_TERNARY_DEFAULT
#define VACOPTVALUE_DISABLED VACOPT_TERNARY_DISABLED
#define VACOPTVALUE_ENABLED VACOPT_TERNARY_ENABLED
#define CopyFromState CopyState
#define BeginCopyFrom_compat(a, b, c, d, e, f, g, h) BeginCopyFrom(a, b, d, e, f, g, h)
#define standard_ProcessUtility_compat(a, b, c, d, e, f, g, h) \
	standard_ProcessUtility(a, b, d, e, f, g, h)
#define ProcessUtility_compat(a, b, c, d, e, f, g, h) ProcessUtility(a, b, d, e, f, g, h)
#define PrevProcessUtility_compat(a, b, c, d, e, f, g, h) \
	PrevProcessUtility(a, b, d, e, f, g, h)
#define COPY_FRONTEND COPY_NEW_FE
#define SetTuplestoreDestReceiverParams_compat(a, b, c, d, e, f) \
	SetTuplestoreDestReceiverParams(a, b, c, d)
#define pgproc_statusflags_compat(pgproc) \
	((&ProcGlobal->allPgXact[(pgproc)->pgprocno])->vacuumFlags)
#define get_partition_parent_compat(a, b) get_partition_parent(a)
#define RelationGetPartitionDesc_compat(a, b) RelationGetPartitionDesc(a)
#define PQ_LARGE_MESSAGE_LIMIT 0
#define make_simple_restrictinfo_compat(a, b) make_simple_restrictinfo(b)
#define pull_varnos_compat(a, b) pull_varnos(b)
#define ROLE_PG_READ_ALL_STATS DEFAULT_ROLE_READ_ALL_STATS
#endif

#define SetListCellPtr(a, b) ((a)->ptr_value = (b))
#define RangeTableEntryFromNSItem(a) ((a)->p_rte)
#define fcGetArgValue(fc, n) ((fc)->args[n].value)
#define fcGetArgNull(fc, n) ((fc)->args[n].isnull)
#define fcSetArgExt(fc, n, val, is_null) \
	(((fc)->args[n].isnull = (is_null)), ((fc)->args[n].value = (val)))
#define fcSetArg(fc, n, value) fcSetArgExt(fc, n, value, false)
#define fcSetArgNull(fc, n) fcSetArgExt(fc, n, (Datum) 0, true)

#endif   /* PG_VERSION_COMPAT_H */
