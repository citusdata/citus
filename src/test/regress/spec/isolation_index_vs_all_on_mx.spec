#include "isolation_mx_common.include.spec"

setup
{
	CREATE TABLE dist_table(id int, data int);
	SELECT create_distributed_table('dist_table', 'id');
	COPY dist_table FROM PROGRAM 'echo 1, 10 && echo 2, 20 && echo 3, 30 && echo 4, 40 && echo 5, 50' WITH CSV;
	SELECT citus_set_coordinator_host('localhost', 57636);
}

teardown
{
	DROP TABLE IF EXISTS dist_table CASCADE;
	SELECT citus_remove_node('localhost', 57636);
}

session "w1"

step "w1-start-session-level-connection"
{
	SELECT start_session_level_connection_to_node('localhost', 57637);
}

step "w1-begin-on-worker"
{
	SELECT run_commands_on_session_level_connection_to_node('BEGIN');
}

step "w1-create-named-index"
{
	SELECT run_commands_on_session_level_connection_to_node('CREATE INDEX dist_table_index ON dist_table(id)');
}

step "w1-create-unnamed-index"
{
	SELECT run_commands_on_session_level_connection_to_node('CREATE INDEX ON dist_table(id,data)');
}

step "w1-reindex"
{
	SELECT run_commands_on_session_level_connection_to_node('REINDEX INDEX dist_table_index');
}

step "w1-delete"
{
	SELECT run_commands_on_session_level_connection_to_node('DELETE FROM dist_table');
}

step "w1-commit-worker"
{
	SELECT run_commands_on_session_level_connection_to_node('COMMIT');
}

step "w1-stop-connection"
{
	SELECT stop_session_level_connection_to_node();
}


session "w2"

step "w2-start-session-level-connection"
{
	SELECT start_session_level_connection_to_node('localhost', 57638);
}

step "w2-begin-on-worker"
{
	SELECT run_commands_on_session_level_connection_to_node('BEGIN');
}

step "w2-create-named-index"
{
	SELECT run_commands_on_session_level_connection_to_node('CREATE INDEX dist_table_index_2 ON dist_table(id)');
}

step "w2-create-unnamed-index"
{
	SELECT run_commands_on_session_level_connection_to_node('CREATE INDEX ON dist_table(id,data)');
}

step "w2-commit-worker"
{
	SELECT run_commands_on_session_level_connection_to_node('COMMIT');
}

step "w2-stop-connection"
{
	SELECT stop_session_level_connection_to_node();
}

session "coord"

step "coord-begin"
{
	BEGIN;
}

step "coord-rollback"
{
	ROLLBACK;
}

step "coord-print-index-count"
{
	SELECT
		result
	FROM
		run_command_on_placements('dist_table', 'select count(*) from pg_indexes WHERE tablename = ''%s''')
	ORDER BY
		nodeport;
}

step "coord-take-lock"
{
	LOCK TABLE dist_table IN ACCESS EXCLUSIVE MODE;
}

session "deadlock-checker"

step "deadlock-checker-call"
{
  SELECT check_distributed_deadlocks();
}

permutation "w1-start-session-level-connection" "w1-begin-on-worker" // open transaction on worker 1
			"w2-start-session-level-connection" "w2-begin-on-worker" // open transaction on worker 2
			"w1-create-named-index" "w2-create-named-index"			 // create indexes with unique names on workers
			"w1-commit-worker" "w2-commit-worker"					 // commit transactions on workers
			"w1-stop-connection" "w2-stop-connection"				 // close connections to workers
			"coord-print-index-count"					 			 // show indexes on coordinator

permutation "w1-start-session-level-connection" "w1-begin-on-worker" // open transaction on worker 1
			"w2-start-session-level-connection" "w2-begin-on-worker" // open transaction on worker 2
			"w1-create-unnamed-index" "w2-create-unnamed-index"		 // create unnamed indexes on workers
			"w1-commit-worker"										 // commit transactions on worker 1 and see error on worker 2 due to name clash
			"w1-stop-connection" "w2-stop-connection"				 // close connections to workers
			"coord-print-index-count"					 			 // show indexes on coordinator

// the following permutation is expected to fail with a distributed deadlock
permutation "w1-start-session-level-connection"				 		 // start session on worker 1 only
			"w1-create-named-index"									 // create index on worker 1
			"w1-begin-on-worker"									 // open transaction block on worker 1
			"w1-delete"												 // delete from table on worker 1
			"coord-begin"											 // open transaction on coordinator to test distributed deadlock
			"coord-take-lock"										 // take ACCESS EXCLUSIVE lock on table on coordinator, get blocked on worker 1
			"w1-reindex"											 // reindex on worker that will acquire ACCESS EXCLUSIVE lock on table, create distributed deadlock
			"deadlock-checker-call"									 // check that distributed deadlock is detected properly
			"coord-rollback"										 // rollback transaction on coordinator to unblock
			"w1-commit-worker"										 // commit transaction on worker 1
			"w1-stop-connection"									 // close connection to worker 1
