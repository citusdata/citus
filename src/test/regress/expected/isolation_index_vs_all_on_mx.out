Parsed test spec with 4 sessions

starting permutation: w1-start-session-level-connection w1-begin-on-worker w2-start-session-level-connection w2-begin-on-worker w1-create-named-index w2-create-named-index w1-commit-worker w2-commit-worker w1-stop-connection w2-stop-connection coord-print-index-count
citus_set_coordinator_host
---------------------------------------------------------------------

(1 row)

step w1-start-session-level-connection:
 SELECT start_session_level_connection_to_node('localhost', 57637);

start_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-begin-on-worker:
 SELECT run_commands_on_session_level_connection_to_node('BEGIN');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-start-session-level-connection:
 SELECT start_session_level_connection_to_node('localhost', 57638);

start_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-begin-on-worker:
 SELECT run_commands_on_session_level_connection_to_node('BEGIN');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-create-named-index:
 SELECT run_commands_on_session_level_connection_to_node('CREATE INDEX dist_table_index ON dist_table(id)');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-create-named-index:
 SELECT run_commands_on_session_level_connection_to_node('CREATE INDEX dist_table_index_2 ON dist_table(id)');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-commit-worker:
 SELECT run_commands_on_session_level_connection_to_node('COMMIT');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-commit-worker:
 SELECT run_commands_on_session_level_connection_to_node('COMMIT');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-stop-connection:
 SELECT stop_session_level_connection_to_node();

stop_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-stop-connection:
 SELECT stop_session_level_connection_to_node();

stop_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step coord-print-index-count:
 SELECT
  result
 FROM
  run_command_on_placements('dist_table', 'select count(*) from pg_indexes WHERE tablename = ''%s''')
 ORDER BY
  nodeport;

result
---------------------------------------------------------------------
     2
     2
     2
     2
(4 rows)

citus_remove_node
---------------------------------------------------------------------

(1 row)


starting permutation: w1-start-session-level-connection w1-begin-on-worker w2-start-session-level-connection w2-begin-on-worker w1-create-unnamed-index w2-create-unnamed-index w1-commit-worker w1-stop-connection w2-stop-connection coord-print-index-count
citus_set_coordinator_host
---------------------------------------------------------------------

(1 row)

step w1-start-session-level-connection:
 SELECT start_session_level_connection_to_node('localhost', 57637);

start_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-begin-on-worker:
 SELECT run_commands_on_session_level_connection_to_node('BEGIN');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-start-session-level-connection:
 SELECT start_session_level_connection_to_node('localhost', 57638);

start_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-begin-on-worker:
 SELECT run_commands_on_session_level_connection_to_node('BEGIN');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-create-unnamed-index:
 SELECT run_commands_on_session_level_connection_to_node('CREATE INDEX ON dist_table(id,data)');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-create-unnamed-index:
 SELECT run_commands_on_session_level_connection_to_node('CREATE INDEX ON dist_table(id,data)');
 <waiting ...>
step w1-commit-worker: 
 SELECT run_commands_on_session_level_connection_to_node('COMMIT');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-create-unnamed-index: <... completed>
ERROR:  duplicate key value violates unique constraint "pg_class_relname_nsp_index"
step w1-stop-connection:
 SELECT stop_session_level_connection_to_node();

stop_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w2-stop-connection:
 SELECT stop_session_level_connection_to_node();

stop_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step coord-print-index-count:
 SELECT
  result
 FROM
  run_command_on_placements('dist_table', 'select count(*) from pg_indexes WHERE tablename = ''%s''')
 ORDER BY
  nodeport;

result
---------------------------------------------------------------------
     1
     1
     1
     1
(4 rows)

citus_remove_node
---------------------------------------------------------------------

(1 row)


starting permutation: w1-start-session-level-connection w1-create-named-index w1-begin-on-worker w1-delete coord-begin coord-take-lock w1-reindex deadlock-checker-call coord-rollback w1-commit-worker w1-stop-connection
citus_set_coordinator_host
---------------------------------------------------------------------

(1 row)

step w1-start-session-level-connection:
 SELECT start_session_level_connection_to_node('localhost', 57637);

start_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-create-named-index:
 SELECT run_commands_on_session_level_connection_to_node('CREATE INDEX dist_table_index ON dist_table(id)');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-begin-on-worker:
 SELECT run_commands_on_session_level_connection_to_node('BEGIN');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-delete:
 SELECT run_commands_on_session_level_connection_to_node('DELETE FROM dist_table');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step coord-begin:
 BEGIN;

step coord-take-lock:
 LOCK TABLE dist_table IN ACCESS EXCLUSIVE MODE;
 <waiting ...>
step w1-reindex: 
 SELECT run_commands_on_session_level_connection_to_node('REINDEX INDEX dist_table_index');
 <waiting ...>
step deadlock-checker-call: 
  SELECT check_distributed_deadlocks();

check_distributed_deadlocks
---------------------------------------------------------------------
t
(1 row)

step coord-take-lock: <... completed>
ERROR:  canceling the transaction since it was involved in a distributed deadlock
step w1-reindex: <... completed>
run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step coord-rollback:
 ROLLBACK;

step w1-commit-worker:
 SELECT run_commands_on_session_level_connection_to_node('COMMIT');

run_commands_on_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

step w1-stop-connection:
 SELECT stop_session_level_connection_to_node();

stop_session_level_connection_to_node
---------------------------------------------------------------------

(1 row)

citus_remove_node
---------------------------------------------------------------------

(1 row)

